src/plugins/mer/mersettings.cpp

```diff
@@ -71,6 +102,124 @@ MerSettings *MerSettings::instance()
+    return FileName::fromString(QFileInfo(globalSettings->fileName()).absolutePath()
+                                + QLatin1String(MER_DEVICE_MODELS_FILENAME));
+}
+
+FileName MerSettings::deviceModelsFileName()
+{
+    const QFileInfo settingsLocation(PluginManager::settings()->fileName());
+    return FileName::fromString(settingsLocation.absolutePath() + QLatin1String(MER_DEVICE_MODELS_FILENAME));
+}
+
+MerEmulatorDeviceModel::Map MerSettings::deviceModels(EmulatorDeviceModelType type)
+{
+    Q_ASSERT(s_instance);
+
+    if (type == EmulatorDeviceModelAll)
+        return s_instance->m_deviceModels;
+
+    const std::map<QString, MerEmulatorDeviceModel> filteredModels
+            = Utils::filtered(s_instance->m_deviceModels.toStdMap(),
+                              [type](const auto &nameValuePair) {
+        return type == EmulatorDeviceModelSdkProvided
+                ? nameValuePair.second.isSdkProvided()
+                : !nameValuePair.second.isSdkProvided();
+    });
+
+    return MerEmulatorDeviceModel::Map(filteredModels);
+}
+
+MerEmulatorDeviceModel::Map MerSettings::deviceModelsRead(const Utils::FileName &fileName)
+{
+    MerEmulatorDeviceModel::Map result;
+    const bool isSdkProvided = (fileName == globalDeviceModelsFileName());
+
+    //! \todo Does not support multiple (different) emulators (not supported at other places anyway).
+    PersistentSettingsReader reader;
+    if (!reader.load(fileName))
+        return result;
+
+    const QVariantMap data = reader.restoreValues();
+
+    const int version = data.value(QLatin1String(MER_DEVICE_MODELS_FILE_VERSION_KEY), 0).toInt();
+    if (version < 1) {
+        qWarning() << "Invalid configuration version: " << version;
+        return result;
+    }
+
+    const int count = data.value(QLatin1String(MER_DEVICE_MODELS_COUNT_KEY), 0).toInt();
+    for (int i = 0; i < count; ++i) {
+        const QString key = QString::fromLatin1(MER_DEVICE_MODELS_DATA_KEY) + QString::number(i);
+        if (!data.contains(key))
+            break;
+
+        const QVariantMap deviceModelData = data.value(key).toMap();
+        MerEmulatorDeviceModel deviceModel(isSdkProvided);
+        deviceModel.fromMap(deviceModelData);
+
+        result.insert(deviceModel.name(), deviceModel);
+    }
+
+    return result;
+}
+
+bool MerSettings::isDeviceModelStored(const MerEmulatorDeviceModel &model)
+{
+    Q_ASSERT(s_instance);
+
+    if (!s_instance->m_deviceModels.contains(model.name()))
+        return false;
+
+    return s_instance->m_deviceModels.value(model.name()) == model;
+}
+
+void MerSettings::setDeviceModels(const MerEmulatorDeviceModel::Map &deviceModels)
+{
+    Q_ASSERT(s_instance);
+
+    if (s_instance->m_deviceModels == deviceModels)
+        return;
+
+    // TODO: use std::transform_reduce (C++17)
+    const MerEmulatorDeviceModel::Map modelsChanged {
+        Utils::filtered(s_instance->m_deviceModels.toStdMap(),
```

Comment:
It took me a while to realize why this kind of initialization is used here. It's not immediately clear that it is needed for the type conversion from std::map to QMap. I would do it here similarly to what you did in deviceModels() - assign to std::map, then convert.

Reply:
Done. Also added reference to name.